src/Bicep.Core/Parsing/Parser.cs

```diff
@@ -38,7 +39,13 @@ public override ProgramSyntax Program()
 
             var endOfFile = reader.Read();
 
-            return new ProgramSyntax(declarationsOrTokens, endOfFile, this.lexerDiagnostics);
+            var parsingErrorTree = new DiagnosticTree();
+            var parsingErrorVisitor = new ParseDiagnosticsVisitor(parsingErrorTree);
+            var programSyntax = new ProgramSyntax(declarationsOrTokens, endOfFile, this.LexingErrorLookup, parsingErrorTree);
+
+            parsingErrorVisitor.Visit(programSyntax);
```

Comment:
Ideally, we should collect parsing diagnostics while parsing to avoid visiting the whole program syntax tree once to collect parsing errors. The thing that makes it challenging is that when `WithRecovery` and `ThrowIfSkip` are used together, one or more `SkippedTriviaSyntax` nodes with the same zero length span may get dropped, and a new `SyntaxTriviaSyntax` with the same span will be added. As a result, we need to either delete some diagnostics from the parsing error tree or keep a snapshot of the tree inside `ThrowIfSkip` and restore the tree to the snapshot if necessary. I tried both approaches, but each of them caused a few tests to fail, which are hard to debug, and I feel like it's not worth the time to go that far. This could be something to implement in the future if anyone is interested.